File: new/704_binary_search_iterative.cpp
class Solution {
public:
    int search(vector<int>& nums, int target) {
        // idea: binary search using iteration
        // use while loop and left and right pointers to do binary search
        int left = 0;
        int right = nums.size() - 1;

        while (left <= right) {
            int middle = left + (right - left) / 2;
            if (nums[middle] < target) {
                left = middle + 1;
            } else if (nums[middle] > target) {
                right = middle - 1;
            } else {
                return middle;
            }
        }

        return -1;
    }
    // time: O(logn)
    // space: O(1)
};

File: new/python/704_binary_search_iterative.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        # idea: binary search using iteration
        # use while loop and left and right pointers to do binary search
        left = 0
        right = len(nums) - 1

        # should be left <= right because we might have an array of length = 1
        while left <= right:
            middle = left + (right - left) // 2
            if nums[middle] == target:
                return middle
            elif nums[middle] < target:
                left = middle + 1
            else:
                right = middle - 1

        return -1
        # time: O(logn)
        # space: O(1)
